### ISSUE TITLE
Error: BuildStream(StreamConfigNotSupported) with running example

### ISSUE THREAD

**Issue by jBernavaPrah** [10buttons/awedio#3]
Hi! 

I'm trying to run the readme example, on a clean project, but on the audio::start(), I receive this error: 

`Error: BuildStream(StreamConfigNotSupported)` 

Would you already have a solution or a hint for this error?

ps. I'm using a Mac.

Thanks!

**Comment by benhansen-io** [comment]
`audio::start` uses the cpal backend.

It attempts to use the default configurations for the default output device.

I would try constructing your own CpalBackend using `CpalBackend::new` with reasonable values (e.g. 2 channels, 44100 samples...) and see if that works.

I'll test audio::start the next time I have access to a Mac computer too.

**Comment by albinekb** [comment]
Hello! I ran in to this issue on macOS aswell, it seems CoreAudio + cpal only supports f32 sample format for outputs.

Changing the data_callback to use f32, makes the examples work as expected: 


```diff --git a/src/backends/cpal_backend.rs b/src/backends/cpal_backend.rs
index 9272de1..36ae863 100644
--- a/src/backends/cpal_backend.rs
+++ b/src/backends/cpal_backend.rs
@@ -4,7 +4,7 @@
 use crate::{manager::BackendSource, manager::Manager, Sound};
 use cpal::{
     traits::{DeviceTrait, HostTrait, StreamTrait},
-    BuildStreamError, PlayStreamError, StreamError,
+    BuildStreamError, PlayStreamError, Sample, StreamError,
 };
 use std::error::Error;
 
@@ -98,7 +98,7 @@ impl CpalBackend {
             panic!("expected MetadataChanged event")
         };
         let channel_count = self.channel_count;
-        let data_callback = move |buffer: &mut [i16], _info: &cpal::OutputCallbackInfo| {
+        let data_callback = move |buffer: &mut [f32], _info: &cpal::OutputCallbackInfo| {
             assert!(buffer.len() % channel_count as usize == 0);
             renderer.on_start_of_batch();
 
@@ -107,14 +107,14 @@ impl CpalBackend {
                     .next_sample()
                     .expect("renderer should never return an Error");
                 match sample {
-                    crate::NextSample::Sample(s) => s,
+                    crate::NextSample::Sample(s) => f32::from_sample(s),
                     crate::NextSample::MetadataChanged => {
                         unreachable!("we never change metadata mid-batch")
                     }
                     // TODO implement pausing
-                    crate::NextSample::Paused => 0,
+                    crate::NextSample::Paused => 0.0,
                     // TODO implement finishing
-                    crate::NextSample::Finished => 0,
+                    crate::NextSample::Finished => 0.0,
                 }
             });
         };
```

         
I don't know if it causes other issues, so far so good!

In [cpal examples](https://github.com/RustAudio/cpal/blob/0da7ae1d426d471cf352550d0048bac91c8bac38/examples/record_wav.rs#L164-L177), they use `Sample::from_sample` to convert between different sample formats. I'm not that familiar with low-level audio, is there some drawback of using `i16` internally and adding an option/default to convert the i16 samples into the supported type? Other than the extra work to convert back and forth (and lost precision?) 

I guess implementing generic sample type trough-out the lib would be a big task since this lib is focused on i16/low end. 





__Update:__ I had a chat with Claude and got help with implementing a quick generic data_callback like the one in cpal example! I can open a PR if you're OK with this workaround for now.

https://github.com/albinekb/awedio/commit/9b6fc8ce6b8fd79a453526435c09cfeab87850dd

**Comment by benhansen-io** [comment]
Thanks @albinekb! A generic function that takes in any sample type and converts to an i16 sounds ideal. Please submit a pull request if you are willing to.

I definitely want to keep Sample inside Awedio to i16 since a generic parameter there complicates the API. From what I have read i16 is more than enough for human hearing range and if someone wants to do DSP or other processing that is better done with an f32, its very cheap to convert back and forth.

**Comment by benhansen-io** [comment]
The fix by @albinekb is in release v0.4.1. @jBernavaPrah if you still are having issues please reopen or start a new issue.
